### PR TITLE
8272315: Improve assert_different_registers

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this improvement to assert_different_register.

It reduces ~250 lines of repetitive code to ~10 lines, and improves the
error message on failure by printing a duplicated registers name.

Unfortunately, this isn't enough to eliminate AbstractRegister[Impl].
Although it seems to no longer be needed for the Oracle-supported platforms,
the linux-arm/ppc64le/s390 platforms all use the value() function provided
by AbstractRegisterImpl.

Testing:
tier1 on Oracle supported platforms.
built linux-arm/ppc64le/s390.

Locally changed a call to have a duplicate register and verified the error
message had a useful register name.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272315](https://bugs.openjdk.java.net/browse/JDK-8272315): Improve assert_different_registers

### Reviewers
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to 839a5bb887e1c6f43c93af943e3e899d4c6700e3
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 839a5bb887e1c6f43c93af943e3e899d4c6700e3
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 839a5bb887e1c6f43c93af943e3e899d4c6700e3

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5083/head:pull/5083` \
`$ git checkout pull/5083`

Update a local copy of the PR: \
`$ git checkout pull/5083` \
`$ git pull https://git.openjdk.java.net/jdk pull/5083/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5083`

View PR using the GUI difftool: \
`$ git pr show -t 5083`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5083.diff">https://git.openjdk.java.net/jdk/pull/5083.diff</a>

</details>

### CODE REVIEW

**Review by adinn — APPROVED** [review]
Looks good.

**Review by iwanowww — APPROVED** [review]
Looks nice!

File: src/hotspot/share/asm/register.hpp [hunk 1]
```diff
@@ -60,263 +61,18 @@ const type name = ((type)value)
...
-    ", l=" INTPTR_FORMAT "",
-    p2i(a), p2i(b), p2i(c), p2i(d), p2i(e), p2i(f), p2i(g), p2i(h), p2i(i), p2i(j), p2i(k), p2i(l)
-  );
+template<typename R, typename... Rx>
+inline void assert_different_registers(R first_register, Rx... more_registers) {
+#ifdef ASSERT
+  R regs[] = { first_register, more_registers... };
+  // Sort regs, so any equal entries are adjacent.
+  struct Compare { ptrdiff_t operator()(R x, R y) const { return x - y; } };
+  QuickSort::sort(regs, ARRAY_SIZE(regs), Compare(), false);
```
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:70]: Sorting is kind of overkill.  Maybe I should have just use the obvious nested loop?  What do reviewers think?

Comment (kimbarrett) [reply src/hotspot/share/asm/register.hpp:70]: Answering my own question, sorting imposes additional requirements on the type that might require adjustment for JDK-8270140.  So I'm inclined to switch to the nested loop that only requires equality testing as before.

Comment (adinn) [reply src/hotspot/share/asm/register.hpp:70]: Well, a nested loop is order n^2 but n is always going to be pretty small (the old code allowed for 12 but cases with n > 6 are rare). So I see no great perf problem with using a nested loop.

Comment (adinn) [reply src/hotspot/share/asm/register.hpp:70]: ... bearing in mind that quicksort is n.log(n) at best.

Comment (iwanowww) [reply src/hotspot/share/asm/register.hpp:70]: > Sorting is kind of overkill. Maybe I should have just use the obvious nested loop? What do reviewers think?

No strong preference on my side. I'm fine with it either way.

**Comment by dean-long** [comment]
Isn't this macro only used for integer registers?  I have an alternative implementation for arm somewhere that uses RegisterSet, but you can just use an array of bytes or bits and check for duplicates as you go in O(n) time, right?

**Comment by kimbarrett** [comment]
> Isn't this macro only used for integer registers? I have an alternative implementation for arm somewhere that uses RegisterSet, but you can just use an array of bytes or bits and check for duplicates as you go in O(n) time, right?

Probably, but that would require knowing the value range or dynamic allocation of the vector.  And this is debug-only code, so keeping it simple would be nice.  I don't think the N^2 behavior (which is what the old code had too) is a problem for the values of N we're dealing with.

File: src/hotspot/share/asm/register.hpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/asm/register.hpp]: Updated to use nested loops.

**Review by TobiHartmann — APPROVED** [review]
Nice cleanup! Looks good to me.

**Review by adinn — APPROVED** [review]
Still good

**Review by iwanowww — APPROVED** [review]

**Comment by kimbarrett** [comment]
Thanks @TobiHartmann , @adinn , and @iwanowww for reviews.

**Pull request closed without merging**